src/v/crypto/crypto.cc

```diff
@@ -46,6 +51,80 @@ std::ostream& operator<<(std::ostream& os, format_type type) {
 }
 
 namespace internal {
+// This function returns true if the provided digest type is expected to use the
+// default provider regardless if the application is in FIPS mode or not.  The
+// use of said digests are not necessarily disallowed, as long as the operations
+// being performed are not security relevant (e.g. performing a checksum rather
+// than an HMAC)
+bool uses_default_provider(digest_type type) {
+    switch (type) {
+    case digest_type::MD5:
+        return true;
+    case digest_type::SHA256:
+    case digest_type::SHA512:
+        return false;
+    }
+}
+bool uses_default_provider(const EVP_MD* md) {
+    struct callback_func_data {
+        bool digest_type_set = false;
+        // type is not valid unless digest_type_set is true
+        digest_type type = digest_type::MD5;
+    } data;
+
+    const auto fn = [](const char* name, void* vdata) {
+        auto data = static_cast<callback_func_data*>(vdata);
+        if (!data->digest_type_set) {
+            std::string name_str{name};
+            if (name_str == "MD5") {
+                data->type = digest_type::MD5;
+                data->digest_type_set = true;
+            } else if (name_str == "SHA256") {
+                data->type = digest_type::SHA256;
+                data->digest_type_set = true;
+            } else if (name_str == "SHA512") {
+                data->type = digest_type::SHA512;
+                data->digest_type_set = true;
+            }
+        }
+    };
+
+    EVP_MD_names_do_all(md, fn, &data);
```

Comment:
where can i learn more about lambda's being able to be passed like this i didn't realize they'd have the same calling conventions for C too?

Reply:
Appears to be an implicit conversion: https://en.cppreference.com/w/cpp/language/lambda (section `ClosureType::operator ret(*)(params)()`). Pretty cool 🙂